tests/net/socket/reuseaddr_reuseport/src/main.c

```diff
@@ -0,0 +1,1011 @@
+{
+	test_reuseaddr_unspecified_specified_common(AF_INET6,
+						    TEST_IPV6_ANY_ADDR,
+						    TEST_IPV6_ANY_ADDR,
+						    SHOULD_FAIL);
+}
+
+
+static void test_reuseaddr_tcp_listening_common(sa_family_t family,
+						char const *first_ip,
+						char const *second_ip)
+{
+	int server_sock1 = -1;
+	int server_sock2 = -1;
+
+	struct sockaddr bind_addr1;
+	struct sockaddr bind_addr2;
+
+	/* Create the sockets */
+	prepare_sock_tcp(family, first_ip, LOCAL_PORT, &server_sock1, &bind_addr1);
+	prepare_sock_tcp(family, second_ip, LOCAL_PORT, &server_sock2, &bind_addr2);
+
+	/* Bind the first socket */
+	test_bind_success(server_sock1, &bind_addr1, sizeof(bind_addr1));
+
+	/* Set the first socket to LISTEN state */
+	test_listen(server_sock1);
+
+	/* Enable SO_REUSEADDR option for the second socket */
+	test_enable_reuseaddr(server_sock2);
+
+	/* Try to bind the second socket, should fail */
+	test_bind_fail(server_sock2, (struct sockaddr *) &bind_addr2, sizeof(bind_addr2));
+
+	close(server_sock1);
+	close(server_sock2);
+
+	test_context_cleanup();
+}
+
+ZTEST_USER(socket_reuseaddr_test_suite, test_ipv4_tcp_unspecified_listening)
+{
+	test_reuseaddr_tcp_listening_common(AF_INET,
+					    TEST_IPV4_ANY_ADDR,
+					    TEST_MY_IPV4_ADDR);
+}
+
+ZTEST_USER(socket_reuseaddr_test_suite, test_ipv6_tcp_unspecified_listening)
+{
+	test_reuseaddr_tcp_listening_common(AF_INET6,
+					    TEST_IPV6_ANY_ADDR,
+					    TEST_MY_IPV6_ADDR);
+}
+
+ZTEST_USER(socket_reuseaddr_test_suite, test_ipv4_tcp_specified_listening)
+{
+	test_reuseaddr_tcp_listening_common(AF_INET,
+					    TEST_MY_IPV4_ADDR,
+					    TEST_IPV4_ANY_ADDR);
+}
+
+ZTEST_USER(socket_reuseaddr_test_suite, test_ipv6_tcp_specified_listening)
+{
+	test_reuseaddr_tcp_listening_common(AF_INET6,
+					    TEST_MY_IPV6_ADDR,
+					    TEST_IPV6_ANY_ADDR);
+}
+
+
+static void test_reuseaddr_tcp_tcp_time_wait_common(sa_family_t family,
+						    char const *first_ip,
+						    char const *second_ip)
+{
+	int server_sock = -1;
+	int client_sock = -1;
+	int accept_sock = -1;
+
+	struct sockaddr bind_addr;
+	struct sockaddr conn_addr;
+
+	struct sockaddr accept_addr;
+	socklen_t accept_addrlen = sizeof(accept_addr);
+
+	prepare_sock_tcp(family, first_ip, LOCAL_PORT, &server_sock, &bind_addr);
+	prepare_sock_tcp(family, second_ip, LOCAL_PORT, &client_sock, &conn_addr);
+
+	/* Bind the server socket */
+	test_bind_success(server_sock, (struct sockaddr *) &bind_addr, sizeof(bind_addr));
+
+	/* Start listening on the server socket */
+	test_listen(server_sock);
+
+	/* Connect the client */
+	test_connect_success(client_sock, &conn_addr, sizeof(conn_addr));
+
+	/* Accept the client */
+	accept_sock = test_accept(server_sock, &accept_addr, &accept_addrlen);
+
+	/* Close the server socket */
+	close(server_sock);
+
+	/* Close the accepted socket */
+	close(accept_sock);
+
+	/* Wait a short time for the accept socket to enter TIME_WAIT state*/
+	k_msleep(50);
```

Comment:
This might cause some issues in the future if some timing changes etc. Anyway, we can cross that bridge when we get there.